library/src/main/java/com/google/android/exoplayer/extractor/ts/Mpeg2Reader.java

```diff
@@ -0,0 +1,189 @@
+    private boolean isSync;
+
+    private int width;
+    private int height;
+
+    private long timeUs;
+    private long samplePosition;
+
+    private boolean foundFirstSample;
+    private long totalBytesWritten;
+
+    public Mpeg2Reader(TrackOutput output) {
+        super(output);
+    }
+
+    @Override
+    public void seek() {
+        seqHeaderFound = false;
+        samplePosition = 0;
+        foundFirstSample = false;
+        totalBytesWritten = 0;
+    }
+
+    @Override
+    public void consume(ParsableByteArray data, long pesTimeUs, boolean startOfPacket) {
+        while (data.bytesLeft() > 0) {
+            int frameAtOffset = 0;
+            int offset = data.getPosition();
+            int limit = data.limit();
+            byte[] dataArray = data.data;
+            // Append the data to the buffer.
+            totalBytesWritten += data.bytesLeft();
+            output.sampleData(data, data.bytesLeft());
+
+            if (!seqHeaderFound && skipToNextSync(dataArray, offset, limit)) {
+                if (seqHeaderFound) {
+                    flags = C.SAMPLE_FLAG_SYNC;
+                    foundFirstSample = true;
+                    timeUs = pesTimeUs;
+                    findSampleSizeAndPrepareMediaFormat(dataArray, offset); // Seq header has been already been found
+                }
+            }
+
+            if (startOfPacket && seqHeaderFound) {
+
+                frameAtOffset = findValidStartCode(dataArray, offset, limit);
+
+                if (frameAtOffset < limit) {
+                    frameCollected = true;
+                    if (!foundFirstSample)
+                        flags = (isSync == true) ? C.SAMPLE_FLAG_SYNC : 0;
+                    foundFirstSample = false;
+                }
+            }
+
+            if (frameCollected) {
+                int dataWrittenPast = limit - frameAtOffset;
+                int sampleSize = (int) (totalBytesWritten - samplePosition) - dataWrittenPast;
+                if (sampleSize > 0) {
+                    output.sampleMetadata(timeUs, flags, sampleSize, dataWrittenPast, null);
+                    samplePosition = totalBytesWritten - dataWrittenPast;
+                    timeUs = pesTimeUs; // Update the time to new pes
+                }
+                frameCollected = false;
+            }
+        }
+    }
+
+    @Override
+    public void packetFinished() {
+        // Do nothing.
+    }
+
+
+    /**
+     * Locates the next sync word, advancing the position to the byte that immediately follows it.
+     * If a sync word was not located, the position is advanced to the limit.
+     * @param data The buffer whose position should be advanced.
+     * @param startoffset The offset of the data in {@code data}.
+     * @param endoffset The limit (exclusive) of the data in {@code data}.
+     * @return True if a sync word position was found. False otherwise.
+     */
+
+    private boolean skipToNextSync(byte[] data, int startoffset, int endoffset) {
```

Comment:
I don't think the way this class looks for start codes is robust against the code being split half in one byte[] and half in the next. NalUnitUtil.findNalUnit with prefixFlags non-null will do this for you (see the H264Reader as an example).

Note - If you do use this, which I think you should do, then we'll rename it to something without "nal" in the name. Don't do this in this change though.